examples/migration/recursive_fibonacci/task_emulation_layer.h

```diff
@@ -47,22 +47,25 @@ class base_task {
 public:
     base_task() = default;
 
-    base_task(const base_task& t) : m_parent(t.m_parent), m_child_counter(t.m_child_counter.load())
+    base_task(const base_task& t) : m_parent(t.m_parent), m_ref_counter(t.m_ref_counter.load())
     {}
 
     virtual ~base_task() = default;
 
     void operator() () const {
         base_task* parent_snapshot = m_parent;
+        task_type type_snapshot = m_type;
+
```

Comment:
What about creating another m_type of "executing". Store the snapshot and then always set m_type to "executing" before calling this->execute(). On return from this->execute(), you can then detect if that task was recycled as continuation during that execution (if so, you do not decrement parent or delete this). If it remained in "executing" state, it was not recycled during that execution, so you can decrement its parent, remove its self_ref and if it was restored to the "allocated" state, then you can delete it.

Reply:
The problem is that approach is: "once recycled task enters `operator()()` and move state to `executing` the original task will read `m_state` and thinks that it was not recycled". Basically. it is another race, probably we can solve it with monotonic epoch.